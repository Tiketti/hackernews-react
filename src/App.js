import axios from 'axios';

import React, { Component } from 'react';
import './App.css';
import Button from './components/Button';
import Search from './components/Search';
import Table from './components/Table';

const DEFAULT_QUERY = 'redux';
const PATH_BASE = 'https://hn.algolia.com/api/v1';
const PATH_SEARCH = '/search';
const PARAM_SEARCH = 'query=';
const PARAM_PAGE = 'page=';
const PARAM_HITSPERPAGE = 'hitsPerPage=';
const DEFAULT_HITSPERPAGE = '10';

class App extends Component {
  constructor(props) {
    super(props);

    this._isMounted = false;

    this.state = {
      results: null,
      searchKey: '',
      searchTerm: DEFAULT_QUERY,
      error: null,
    };

    this.setSearchTopStories = this.setSearchTopStories.bind(this);
    this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this);
    this.needsToSearchTopStories = this.needsToSearchTopStories.bind(this);
    this.onSearchChange = this.onSearchChange.bind(this);
    this.onSearchSubmit = this.onSearchSubmit.bind(this);
    this.onDismiss = this.onDismiss.bind(this);
  }

  componentDidMount() {
    this._isMounted = true;

    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm });
    this.fetchSearchTopStories(searchTerm);
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  onDismiss(id) {
    const { searchKey, results } = this.state;
    const { hits, page } = results[searchKey];

    const updatedHits = hits.filter(item => item.objectID !== id);
    this.setState({
      results: {
        ...results,
        [searchKey]: { hits: updatedHits, page },
      },
    });
  }

  onSearchChange(event) {
    // eslint-disable-next-line no-console
    console.log(`search term: ${this.state.searchTerm}`);
    this.setState({ searchTerm: event.target.value });
  }

  onSearchSubmit(event) {
    const { searchTerm } = this.state;
    event.preventDefault();

    this.setState({ searchKey: searchTerm });

    if (this.needsToSearchTopStories(searchTerm)) {
      this.fetchSearchTopStories(searchTerm);
    }
  }

  setSearchTopStories(result) {
    const { hits, page } = result;
    const { searchKey, results } = this.state;

    const oldHits = results && results[searchKey]
      ? results[searchKey].hits
      : [];

    const updatedHits = [
      ...oldHits,
      ...hits,
    ];

    this.setState({
      results: {
        ...results,
        [searchKey]: { hits: updatedHits, page },
      },
    });
  }

  needsToSearchTopStories(searchTerm) {
    return !this.state.results[searchTerm];
  }

  async fetchSearchTopStories(searchTerm, page = 0) {
    try {
      const response = await axios(`${PATH_BASE}${PATH_SEARCH}\
?${PARAM_SEARCH}${searchTerm}\
&${PARAM_PAGE}${page}\
&${PARAM_HITSPERPAGE}${DEFAULT_HITSPERPAGE}`);

      if (this._isMounted) {
        this.setSearchTopStories(response.data);
      }
    } catch (error) {
      if (this._isMounted) {
        // eslint-disable-next-line no-console
        console.error(error);
        this.setState({ error });
      }
    }
  }

  render() {
    const {
      searchTerm,
      results,
      searchKey,
      error,
    } = this.state;

    const page = (
      results &&
      results[searchKey] &&
      results[searchKey].page
    ) || 0;

    const list = (
      results &&
      results[searchKey] &&
      results[searchKey].hits
    ) || [];

    return (
      <div className="page">
        <div className="interactions">
          <Search
            value={searchTerm}
            onChange={this.onSearchChange}
            onSubmit={this.onSearchSubmit}
          >
          Search
          </Search>
          <Button
            className="more"
            onClick={() =>
              this.fetchSearchTopStories(searchKey, page + 1)
            }
          >
            More
          </Button>
          {error ?
            <div className="interactions">
              <p>Something went wrong.</p>
            </div>
            :
            <Table
              list={list}
              onDismiss={this.onDismiss}
            />
          }
        </div>
      </div>
    );
  }
}

export default App;
