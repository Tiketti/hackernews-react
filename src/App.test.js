import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import App, { Search, Table } from './App';

jest.mock('axios', () => ({
  get: jest.fn(() => (
    {
      data: {
        hits: '',
        page: '',
      },
    }
  )),
}
));

Enzyme.configure({ adapter: new Adapter() });

describe('App', () => {
  test('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(<App />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Search', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Search>Search</Search>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(<Search>Search</Search>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Table', () => {
  let props;

  beforeAll(() => {
    props = {
      list: [
        { title: '1', author: '1', num_comments: 1, points: 2, objectID: 'y' },
        { title: '2', author: '2', num_comments: 1, points: 2, objectID: 'z' },
      ],
    };
  });

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Table {...props} />, div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(<Table {...props} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('shows two items in list', () => {
    const element = shallow(<Table {...props} />);
    expect(element.find('.table-row').length).toBe(2);
  });
});
