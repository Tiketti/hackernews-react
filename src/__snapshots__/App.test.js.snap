// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a valid snapshot 1`] = `
<div
  className="page"
>
  <div
    className="interactions"
  >
    <form
      onSubmit={[Function]}
    >
      Search
      <input
        onChange={[Function]}
        type="text"
        value="redux"
      />
      <button
        type="submit"
      >
        Search
      </button>
    </form>
    <div>
      Loading ...
    </div>
    <div
      className="table"
    />
  </div>
</div>
`;

exports[`Search has a valid snapshot 1`] = `
<form
  onSubmit={undefined}
>
  Search
  <input
    onChange={undefined}
    type="text"
    value={undefined}
  />
  <button
    type="submit"
  >
    Search
  </button>
</form>
`;

exports[`Table has a valid snapshot 1`] = `
<div
  className="table"
>
  <div
    className="table-row"
  >
    <span
      className="largeColumn"
    >
      <a
        href={undefined}
      >
        1
      </a>
    </span>
    <span
      className="midColumn"
    >
      1
       |
    </span>
    <span
      className="smallColumn"
      title="Number of comments"
    >
      1
       |
    </span>
    <span
      className="smallColumn"
      title="Points"
    >
      2
       
    </span>
    <span
      className="smallColumn"
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      className="largeColumn"
    >
      <a
        href={undefined}
      >
        2
      </a>
    </span>
    <span
      className="midColumn"
    >
      2
       |
    </span>
    <span
      className="smallColumn"
      title="Number of comments"
    >
      1
       |
    </span>
    <span
      className="smallColumn"
      title="Points"
    >
      2
       
    </span>
    <span
      className="smallColumn"
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
</div>
`;
